### PR TITLE
Implement dead key sequences.

### PR DESCRIPTION
**Pull request by Nuclear-Squid**
Hi ! With ergol needing deadkey sequences, I implemented the feature into X-Keyboard. Most of the time it was pretty easy (just needing to let the control flow fall through, instead of early returning), though I’m not sure my way of implementing the keyhint for the dead key on 1dk layer is the cleanest. It ~ seems ~ to be working fine, but if you have a better idea please let me know. (I also updated ergol to version 0.99.2, to test out the double dead key)

### CODE REVIEW

**Comment by fabi1cazenave** [comment]
Thanks for adding the JSON file I forgot to update with my latest patch… shame on me.

**Comment by fabi1cazenave** [comment]
Adding the JSON for Ergo-L 0.99.2 is nice, but please provide the corresponding YAML file as well (or a TOML file and update the Makefile accordingly so both YAML and TOML files are built).

File: src/content.js [hunk 1]
```diff
@@ -155,8 +155,13 @@ export function drawDK(element, keyMap, deadKey) {
...
   const alt1 = deadKey[keyChars[1]];
-  element.querySelector('.level5').textContent = alt0 || '';
-  element.querySelector('.level6').textContent = altUpperChar(alt0, alt1);
+  if (isDeadKey(alt0)) {
+    element.querySelector('.level5').classList.add('deadKey', 'diacritic');
+    element.querySelector('.level5').textContent = alt0[1];
+  } else {
+    element.querySelector('.level5').textContent = alt0 || '';
+    element.querySelector('.level6').textContent = altUpperChar(alt0, alt1);
+  }
```
Comment (fabi1cazenave) [inline src/content.js:164]: That’s pretty close, and it works fine with Ergo-L 0.99.2. Two remarks :
- the `$(.level6)` text content is undetermined when `alt0` is a dead key
- the `deadKey` and `diacritic` classes are never removed

So I’d rather suggest something like this : 

```js
  const lvl5 = element.querySelector('.level5');
  const lvl6 = element.querySelector('.level6');
  if (isDeadKey(alt0)) {
    lvl5.classList.add('deadKey', 'diacritic');
    lvl5.textContent = alt0[1];
  } else {
    lvl5.classList.remove('deadKey', 'diacritic');
    lvl5.textContent = alt0 || '';
  }
  lvl6.textContent = altUpperChar(alt0, alt1);
```

To test this, put a dead key in the `1dk` layer on a key that also has a defined symbol on `1dk`<kbd>Shift</kbd>[letter], like for the comma key here : 
```
  ╭╌╌╌╌╌┰─────┬─────┬─────┬─────┬─────┰─────┬─────┬─────┬─────┬─────┰╌╌╌╌╌┬╌╌╌╌╌╮
  ┆ ~   ┃ ! ¡ │ @ ‘ │ # ’ │ $ ¢ │ % ‰ ┃ ^   │ &   │ *   │ «   │ »   ┃ _ – ┆ + ± ┆
  ┆ `   ┃ 1 „ │ 2 “ │ 3 ” │ 4 £ │ 5 € ┃ 6 ¥ │ 7 ¤ │ 8 § │ 9 ¶ │ 0 ° ┃ / ÷ ┆ = ≠ ┆
  ╰╌╌╌╌╌╂─────┼─────┼─────┼─────┼─────╂─────┼─────┼─────┼─────┼─────╂╌╌╌╌╌┼╌╌╌╌╌┤
  ·     ┃ Q   │ C   │ O   │ P   │ W   ┃ J   │ M   │ D   │**   │ F   ┃ {   ┆ }   ┆
  ·     ┃   â │   ç │   œ │   ô │     ┃     │   µ │   _ │***¨ │     ┃ [   ┆ ]   ┆
  ·     ┠─────┼─────┼─────┼─────┼─────╂─────┼─────┼─────┼─────┼─────╂╌╌╌╌╌┼╌╌╌╌╌┤
  ·     ┃ A   │ S   │ E   │ N   │ ; • ┃ L   │ R   │ T   │ I   │ U   ┃ "   ┆ |   ┆
  ·     ┃   à │   é │   è │   ê │ ,*` ┃     │     │   î │   û │   ù ┃ '   ┆ \   ┆
  ╭╌╌╌╌╌╂─────┼─────┼─────┼─────┼─────╂─────┼─────┼─────┼─────┼─────╂╌╌╌╌╌┴╌╌╌╌╌╯
  ┆ >   ┃ Z   │ X   │ ? ¿ │ V   │ B   ┃ :   │ H   │ G   │ Y   │ K   ┃           ·
  ┆ <   ┃     │   ß │ - — │   ñ │     ┃ . … │     │     │     │     ┃           ·
  ╰╌╌╌╌╌┸─────┴─────┴─────┴─────┴─────┸─────┴─────┴─────┴─────┴─────┚ · · · · · ·
```

File: src/x-keyboard.js [hunk 2]
```diff
@@ -179,7 +179,8 @@ class Keyboard extends HTMLElement {
             span.textContent = '';
           });
       }
-    } else if (this.layout.pendingDK) { // show hints for this dead key
+    }
+    if (this.layout.pendingDK) { // show hints for this dead key
```
Comment (fabi1cazenave) [inline src/x-keyboard.js:183]: Brilliant.

**Review by fabi1cazenave — CHANGES_REQUESTED** [review]
Nice and helpful job, congrats and thanks !

I’ve pointed to a couple changes that would finish the job. Please make sure the Ergo-L YAML is up-to-date as well.

**Comment by Nuclear-Squid** [comment]
< requested-changes>
I found another bug in your suggestion for the `drawDK` function, where if the level6 character is a dead key and needs to be shown, it will be shown as a regular character. To fix that, I rewrote a bit of that function to factor out the "render the character" part into it’s own closure (or whatever a local lambda is called in JS). I think it looks clean, but I’ll let you judge that by yourself.
</ requested-changes>

< extra-and-possibly-not-relevent-bug>
I also took the time to investigate another minor bug, where pressing a deadkey (other than `1dk`), deletes last character of the `input` element. For instance, if I write "Hello" then press `1dk` twice, the 'o' is removed. It seems to be because when a dead key is pressed, the event listener in `demo.js` on line 92 (see below) who hides the hint char (by deleting it) is called twice in very quick succession.

```js
input.addEventListener('input', (event) => {
  if ((event.inputType === 'insertCompositionText'
    || event.inputType === 'insertText') && event.isComposing) {
    event.target.value = event.target.value.slice(0, -event.data.length);
  }
});
```

Logging the events (see screenshot below) shows they are basically the same, except the first event has the `isComposing` field set to `true` and the second one to `false`.

Screenshot_2024-01-04_18-06-00

Removing the last char only when `event.isComposing` is set to `true` fixes that problem, and seems to be the only even we care about, if I understand [MDN’s documentation](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/isComposing) correctly.

```js
input.addEventListener('input', (event) => {
  if ((event.inputType === 'insertCompositionText'
    || event.inputType === 'insertText') && event.isComposing) {
    event.target.value = event.target.value.slice(0, -event.data.length);
  }
});
```

Is there an edge case I missed ? (God knows JS has a lot of those). If you like that fix, do I push another commit onto that branch, or do I open another PR ?
</ extra-and-possibly-not-relevent-bug>

File: src/content.js [hunk 3]
```diff
@@ -149,14 +149,24 @@ export function drawKey(element, keyMap) {
 }
 
 export function drawDK(element, keyMap, deadKey) {
```
Comment (fabi1cazenave) [inline src/content.js:151]: Do we need the `export` keyword here ? (= do we really want to export this function ?)

**Review by fabi1cazenave — COMMENTED** [review]
Everything looks good to me, please clarify the `export` part and we’re ready to merge. :-)

**Comment by fabi1cazenave** [comment]
About the `extra-and-possibly-not-relevent-bug`, I confess I haven’t met it yet. Maybe I’m too slow ? However, if you’re positive about that bug I’d love to see a fix in another PR. This would be an opportunity to simplify `demo.js`, which contains a workaround for a Chrome bug which might have been fixed since.

File: src/content.js (hunk 3 above)
Comment (Nuclear-Squid) [reply src/content.js:151]: I’d say yes, it is used by `x-keyboard.js` : https://github.com/fabi1cazenave/x-keyboard/pull/29/files#diff-6a4b48a52a7703e5dc74b3a02c11978ba66663dda879d004c5767c56f1c26bf8R184-R186

Comment (fabi1cazenave) [reply src/content.js:151]: Duh. Indeed.

**Comment by fabi1cazenave** [comment]
Thanks for this very good work ! Your help is very appreciated.

**Pull request merged**